### PR TITLE
feat(gatsby): first pass at API functions docs

### PR DESCRIPTION
**Pull request by KyleAMathews**
Screenshot_2021-04-27 Gatsby

### CODE REVIEW

File: packages/gatsby/index.d.ts [hunk 1]
```diff
@@ -1529,9 +1529,9 @@ export interface PluginOptionsSchemaArgs {
 type Send<T> = (body: T) => void
 
 /**
- * Gatsby API Function route response
+ * Gatsby Function route response
  */
-export interface GatsbyAPIFunctionResponse extends ServerResponse {
+export interface GatsbyFunctionResponse extends ServerResponse {
```
Comment (KyleAMathews) [inline packages/gatsby/index.d.ts:1534]: part of renaming from `API Functions` to just `Functions`

File: integration-tests/functions/test-helpers.js [hunk 2]
```diff
@@ -220,6 +220,15 @@ export function runTests(env, host) {
       })
     })
 
+    describe(`functions can have custom middleware`, () => {
```
Comment (KyleAMathews) [inline integration-tests/functions/test-helpers.js:223]: Added a test for custom middleware while writing the docs for it

File: docs/docs/how-to/functions/routing.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+## Static Routing
+
+Both top-level and nested routes are supported.
+
+- `src/api/top-level.js` => `/api/top-level`
+- `src/api/directory/foo.js` => `/api/directory/foo`
+
+`index.js` files are routed at their directory path e.g. `src/api/users/index.js` => `/api/users`
+
+## Dynamic Routing
```
Comment (jsumnersmith) [inline docs/docs/how-to/functions/routing.md:16]: Not yet supported in Cloud, right?

File: docs/docs/how-to/functions/getting-started.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
+---
+title: Getting Started
```
Comment (jsumnersmith) [inline docs/docs/how-to/functions/getting-started.md:2]: I feel like this is missing any info about deploying in Cloud. I had a paragraph in G-docs about that.

Comment (KyleAMathews) [reply docs/docs/how-to/functions/getting-started.md:2]: I put the paragraph on the overview page (in the mansion PR https://github.com/gatsby-inc/mansion/pull/11587)

File: docs/docs/how-to/functions/routing.md (hunk 3 above)
Comment (KyleAMathews) [reply docs/docs/how-to/functions/routing.md:16]: not yet but close

Comment (jsumnersmith) [reply docs/docs/how-to/functions/routing.md:16]: Do we need to note that then? I don't think the Fastly piece is being worked on at the moment (maybe next sprint).

Comment (KyleAMathews) [reply docs/docs/how-to/functions/routing.md:16]: Yeah, if we're not going to be ready by docs launch then we should note it. Adding...

**Review by jsumnersmith — APPROVED** [review]
Looks good! Let's get 'em up!

**Pull request merged**